architecture.tex

```diff
@@ -53,58 +55,190 @@
+named functions like $\sin(x)$ and $\log(x)$ as well as undefined functions
+like $f(x)$. Subclasses of \texttt{Function} should define a
+class method \texttt{eval}, which returns values for which the function should
+be automatically evaluated, and \texttt{None} for arguments that shouldn't be
+automatically evaluated.
+
+The behavior of classes in SymPy with various other SymPy functions is defined
+by defining a relevant \verb|_eval_|\texttt{\textit{*}} method on the class. For instance, an
+object can tell SymPy's \texttt{diff} function how to take the derivative of
+itself by defining the \verb|_eval_derivative(self, x)| method. The most
+common \verb|_eval_|\texttt{\textit{*}} methods relate to the assumptions.
+\verb|_eval_is_|\texttt{\textit{assumption}} defines the assumptions for
+\textit{assumption}.
+
+Here is a stripped down version of the gamma function $\Gamma(x)$ from SymPy,
+which evaluates itself on positive integer arguments, has the positive and
+real assumptions defined, can be rewritten in terms of factorial with
+\texttt{gamma(x).rewrite(factorial)}, and can be differentiated.
+\texttt{fdiff} is a convenience method for subclasses of \texttt{Function}.
+\texttt{fdiff} returns the derivative of the function without worrying about
+the chain rule. \texttt{self.func} is used throughout instead of referencing
+\texttt{gamma} explicitly so that potential subclasses of \texttt{gamma} can
+reuse the methods.
 
-Mathematical functions are implemented as Python classes.
-
-%% Pattern matching
-%%
-%%%% three kinds of pattern matching in SymPy core.
-%%%% `strategies` and `unify` modules also worth mentioning
-%%%% (fu trig simplification function).
-%%%% Rewriting rules.
+\begin{verbatim}
+from sympy import Integer, Function, floor, factorial, polygamma
+
+class gamma(Function)
+    @classmethod
+    def eval(cls, arg):
+        if isinstance(arg, Integer) and arg.is_positive:
+            return factorial(arg - 1)
+
+    def _eval_is_real(self):
+        x = self.args[0]
+        # noninteger means real and not integer
+        if x.is_positive or x.is_noninteger:
+            return True
+
+    def _eval_is_positive(self):
+        x = self.args[0]
+        if x.is_positive:
+            return True
+        elif x.is_noninteger:
+            return floor(x).is_even
+
+    def _eval_rewrite_as_factorial(self, z):
+        return factorial(z - 1)
+
+    def fdiff(self, argindex=1):
+        from sympy.core.function import ArgumentIndexError
+        if argindex == 1:
+            return self.func(self.args[0])*polygamma(0, self.args[0])
+        else:
+            raise ArgumentIndexError(self, argindex)
+\end{verbatim}
 
-%% Withheld evaluated (appears frequently on StackOverflow as a question).
+The actual gamma function defined in SymPy has much more implemented than
+this, such as evaluation at rational points, series expansion, and pretty
+printing with the $\Gamma$ symbol.
```

Comment:
I'm not sure how to do that concisely. But I might delete this anyway because technically the printing is defined in the printer, not the `gamma` class, and even if you don't tell the printer anything, it knows how to print greek letters correctly. If we end up adding a section for the printers we can add this back in.